### PR TITLE
[AArch64][GlobalISel] Make G_DUP immediate 32-bits or larger

### PR DESCRIPTION
**Pull request by chuongg3**
G_DUP's immediate operand gets extended in RegBankSelect to allow for better pattern matching in TableGen for #96782

### CODE REVIEW

File: llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp [hunk 1]
```diff
@@ -420,6 +420,25 @@ void AArch64RegisterBankInfo::applyMappingImpl(
...
+    auto ConstMI = MRI.getVRegDef(MI.getOperand(1).getReg());
+    if (ConstMI->getOpcode() == TargetOpcode::G_CONSTANT) {
+      auto CstVal = ConstMI->getOperand(1).getCImm()->getValue();
+      ConstReg =
+          Builder.buildConstant(LLT::scalar(32), CstVal.sext(32)).getReg(0);
+      ConstMI->eraseFromParent();
+    } else {
+      ConstReg = Builder.buildAnyExt(LLT::scalar(32), MI.getOperand(1).getReg())
+                     .getReg(0);
+    }
```
Comment (aemerson) [inline llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp:437]: What if it's an i64 G_CONSTANT?

Comment (chuongg3) [reply llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp:437]: i32 and i64 constants are not changed, only constants that are smaller than 32-bits get extended

File: llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp [hunk 2]
```diff
@@ -774,8 +793,12 @@ AArch64RegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
              (getRegBank(ScalarReg, MRI, TRI) == &AArch64::FPRRegBank ||
               onlyDefinesFP(*ScalarDef, MRI, TRI)))
       OpRegBankIdx = {PMI_FirstFPR, PMI_FirstFPR};
-    else
+    else {
+      if (ScalarTy.getSizeInBits() < 32 &&
+          getRegBank(ScalarReg, MRI, TRI) == &AArch64::GPRRegBank)
+        MappingID = 1;
```
Comment (aemerson) [inline llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp:799]: Can you make `1` here a `const unsigned` with a name like `CustomID` so it's clearer? Also add a comment here so the reader knows it causes `applyMappingImpl` to be called.

**Review by aemerson — APPROVED** [review]
LGTM with suggestions.

File: llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp (hunk 1 above)
Comment (davemgreen) [reply llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp:437]: It might be good to add an assert that the type isn't larger than we expect

**Review by davemgreen — APPROVED** [review]
LGTM

**Comment by DavidSpickett** [comment]
This PR caused some of the test suite build to crash the compiler: https://lab.llvm.org/buildbot/#/builders/125/builds/541

```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang++ -DNDEBUG -O0 -fPIE -w -Werror=date-time -MD -MT Bitcode/simd_ops/CMakeFiles/simd_ops_test_op_cmeq_1056.dir/AArch64_tests/test_op_cmeq_1056.bc.o -MF CMakeFiles/simd_ops_test_op_cmeq_1056.dir/AArch64_tests/test_op_cmeq_1056.bc.o.d -o CMakeFiles/simd_ops_test_op_cmeq_1056.dir/AArch64_tests/test_op_cmeq_1056.bc.o -c /home/tcwg-buildbot/worker/clang-aarch64-global-isel/test/test-suite/Bitcode/simd_ops/AArch64_tests/test_op_cmeq_1056.bc
1.	Code generation
2.	Running pass 'Function Pass Manager' on module '/home/tcwg-buildbot/worker/clang-aarch64-global-isel/test/test-suite/Bitcode/simd_ops/AArch64_tests/test_op_cmeq_1056.bc'.
3.	Running pass 'RegBankSelect' on function '@__test_op'
<...>
 #0 0x0000aaaae0c95714 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x3731714)
 #1 0x0000aaaae0c934d8 llvm::sys::RunSignalHandlers() (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x372f4d8)
 #2 0x0000aaaae0c94a6c llvm::sys::CleanupOnSignal(unsigned long) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x3730a6c)
 #3 0x0000aaaae0c07d3c CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #4 0x0000ffff941e3598 (linux-vdso.so.1+0x598)
 #5 0x0000aaaade912324 llvm::AArch64RegisterBankInfo::getInstrMapping(llvm::MachineInstr const&) const (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x13ae324)
 #6 0x0000aaaae18e3078 llvm::RegBankSelect::assignInstr(llvm::MachineInstr&) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x437f078)
 #7 0x0000aaaae18e34c4 llvm::RegBankSelect::assignRegisterBanks(llvm::MachineFunction&) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x437f4c4)
 #8 0x0000aaaae18e3804 llvm::RegBankSelect::runOnMachineFunction(llvm::MachineFunction&) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x437f804)
 #9 0x0000aaaae01c1db8 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x2c5ddb8)
#10 0x0000aaaae06e6544 llvm::FPPassManager::runOnFunction(llvm::Function&) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x3182544)
#11 0x0000aaaae06ee158 llvm::FPPassManager::runOnModule(llvm::Module&) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x318a158)
#12 0x0000aaaae06e6f30 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x3182f30)
#13 0x0000aaaae0e949d0 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x39309d0)
#14 0x0000aaaae139cfc0 clang::CodeGenAction::ExecuteAction() (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x3e38fc0)
#15 0x0000aaaae15b6264 clang::FrontendAction::Execute() (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x4052264)
#16 0x0000aaaae153a6ec clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x3fd66ec)
#17 0x0000aaaae1692dd4 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x412edd4)
#18 0x0000aaaade7d6cf0 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x1272cf0)
#19 0x0000aaaade7d3c44 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#20 0x0000aaaae13df14c void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::$_0>(long) Job.cpp:0:0
#21 0x0000aaaae0c07a4c llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x36a3a4c)
#22 0x0000aaaae13de730 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x3e7a730)
#23 0x0000aaaae13a5edc clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x3e41edc)
#24 0x0000aaaae13a612c clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x3e4212c)
#25 0x0000aaaae13bf48c clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x3e5b48c)
#26 0x0000aaaade7d3048 clang_main(int, char**, llvm::ToolContext const&) (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x126f048)
#27 0x0000aaaade7e1194 main (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x127d194)
#28 0x0000ffff93d25e10 __libc_start_main /build/glibc-Q8DG8B/glibc-2.31/csu/../csu/libc-start.c:342:3
#29 0x0000aaaade7d1768 _start (/home/tcwg-buildbot/worker/clang-aarch64-global-isel/stage1.install/bin/clang+++0x126d768)
make[2]: Entering directory '/home/tcwg-buildbot/worker/clang-aarch64-global-isel/test/sandbox/build'
```

I'll dig out a reproducer, but you can also compile using the files in llvm-test-suite.

**Comment by DavidSpickett** [comment]
[scal-to-vec1-719b0d.zip](https://github.com/user-attachments/files/16142899/scal-to-vec1-719b0d.zip)

```
$ ./scal-to-vec1-719b0d.sh
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /home/david.spickett/build-llvm-aarch64/bin/clang-19 -cc1 -triple aarch64-unknown-linux-gnu -emit-obj -disable-free -clear-ast-before-backend -main-file-name scal-to-vec1.c -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=non-leaf -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu generic -target-feature +v8a -target-feature +fp-armv8 -target-feature +neon -target-abi aapcs -debugger-tuning=gdb -fdebug-compilation-dir=/home/tcwg-buildbot/worker/clang-aarch64-global-isel/test/sandbox/build/SingleSource/Regression/C/gcc-c-torture/execute -fcoverage-compilation-dir=/home/tcwg-buildbot/worker/clang-aarch64-global-isel/test/sandbox/build/SingleSource/Regression/C/gcc-c-torture/execute -sys-header-deps -D NDEBUG -O0 -Werror=date-time -Wno-implicit-int -Wno-int-conversion -Wno-implicit-function-declaration -w -ferror-limit 19 -fno-signed-char -fgnuc-version=4.2.1 -fskip-odr-check-in-gmf -target-feature +outline-atomics -target-feature -fmv -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -x c scal-to-vec1-719b0d.c
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module 'scal-to-vec1-719b0d.c'.
4.      Running pass 'RegBankSelect' on function '@main'
 #0 0x0000aaaad3808304 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x3736304)
 #1 0x0000aaaad3806104 llvm::sys::RunSignalHandlers() (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x3734104)
 #2 0x0000aaaad3808a2c SignalHandler(int) Signals.cpp:0:0
 #3 0x0000ffffb29015c0 (linux-vdso.so.1+0x5c0)
 #4 0x0000aaaad14801c8 llvm::AArch64RegisterBankInfo::getInstrMapping(llvm::MachineInstr const&) const (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x13ae1c8)
 #5 0x0000aaaad44591ac llvm::RegBankSelect::assignInstr(llvm::MachineInstr&) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x43871ac)
 #6 0x0000aaaad4459608 llvm::RegBankSelect::assignRegisterBanks(llvm::MachineFunction&) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x4387608)
 #7 0x0000aaaad4459948 llvm::RegBankSelect::runOnMachineFunction(llvm::MachineFunction&) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x4387948)
 #8 0x0000aaaad2d346e0 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x2c626e0)
 #9 0x0000aaaad32594b8 llvm::FPPassManager::runOnFunction(llvm::Function&) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x31874b8)
#10 0x0000aaaad3260f44 llvm::FPPassManager::runOnModule(llvm::Module&) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x318ef44)
#11 0x0000aaaad3259ea8 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x3187ea8)
#12 0x0000aaaad3a06170 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x3934170)
#13 0x0000aaaad3f0a8c4 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x3e388c4)
#14 0x0000aaaad58a068c clang::ParseAST(clang::Sema&, bool, bool) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x57ce68c)
#15 0x0000aaaad412ccf4 clang::FrontendAction::Execute() (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x405acf4)
#16 0x0000aaaad40b1308 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x3fdf308)
#17 0x0000aaaad420b3e4 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x41393e4)
#18 0x0000aaaad13433e0 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x12713e0)
#19 0x0000aaaad1340344 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#20 0x0000aaaad133eff8 clang_main(int, char**, llvm::ToolContext const&) (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x126cff8)
#21 0x0000aaaad134d7f0 main (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x127b7f0)
#22 0x0000ffffb243ce10 __libc_start_main /build/glibc-Q8DG8B/glibc-2.31/csu/../csu/libc-start.c:342:3
#23 0x0000aaaad133ddb8 _start (/home/david.spickett/build-llvm-aarch64/bin/clang-19+0x126bdb8)
./scal-to-vec1-719b0d.sh: line 1: 3165123 Segmentation fault      (core dumped) "/home/david.spickett/build-llvm-aarch64/bin/clang-19" "-cc1" "-triple" "aarch64-unknown-linux-gnu" "-emit-obj" "-disable-free" "-clear-ast-before-backend" "-main-file-name" "scal-to-vec1.c" "-mrelocation-model" "pic" "-pic-level" "2" "-pic-is-pie" "-mframe-pointer=non-leaf" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-funwind-tables=2" "-target-cpu" "generic" "-target-feature" "+v8a" "-target-feature" "+fp-armv8" "-target-feature" "+neon" "-target-abi" "aapcs" "-debugger-tuning=gdb" "-fdebug-compilation-dir=/home/tcwg-buildbot/worker/clang-aarch64-global-isel/test/sandbox/build/SingleSource/Regression/C/gcc-c-torture/execute" "-fcoverage-compilation-dir=/home/tcwg-buildbot/worker/clang-aarch64-global-isel/test/sandbox/build/SingleSource/Regression/C/gcc-c-torture/execute" "-sys-header-deps" "-D" "NDEBUG" "-O0" "-Werror=date-time" "-Wno-implicit-int" "-Wno-int-conversion" "-Wno-implicit-function-declaration" "-w" "-ferror-limit" "19" "-fno-signed-char" "-fgnuc-version=4.2.1" "-fskip-odr-check-in-gmf" "-target-feature" "+outline-atomics" "-target-feature" "-fmv" "-faddrsig" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-x" "c" "scal-to-vec1-719b0d.c"
```

**Comment by aemerson** [comment]
@chuongg3 @davemgreen ^

**Comment by davemgreen** [comment]
Thanks for the info @DavidSpickett. Sorry we didn't reply here. @chuong is looking into it!

**Pull request merged**